### ISSUE TITLE
How to map objects when source object has custom method for Getter?

### ISSUE THREAD

**Issue by wonsuc** [mapstruct/mapstruct#2753]
I'm using a specific framework(it's called TeamCenter), I want to use this library to map from that framework's object to my custom DTO.

However, this framework's model class is `ModelObject` and to get values from the object, I have to call method like this => 
`modelObject.getPropertyDisplayableValue("propertyName");`

How can I achieve this with `MapStruct`? 

I created a `Mapper` class like below.  

```
@Mapper
public interface ProjectMapper {
    ProjectMapper INSTANCE = Mappers.getMapper(ProjectMapper.class);

    @Mapping(target = "id", source = "prg0PlanId")
    ProjectDto toDto(ModelObject projectModel) throws NotLoadedException;
}
```

The only working way is to use `expression`.  

```
@Mapper
public interface ProjectMapper {
    ProjectMapper INSTANCE = Mappers.getMapper(ProjectMapper.class);

    @Mapping(target = "id", expression = "java(projectModel.getPropertyDisplayableValue(\"prg0PlanId\")")
    ProjectDto toDto(ModelObject projectModel) throws NotLoadedException;
}
```

But, I think this is not the advantage to use `MapStruct`.  

I want to use something like this.  

```
@Mapper
public interface ProjectMapper {
    ProjectMapper INSTANCE = Mappers.getMapper(ProjectMapper.class);

    @Mapping(target = "id", source = "prg0PlanId", qualifiedByName = "getPropertyDisplayableValue")
    ProjectDto toDto(ModelObject projectModel) throws NotLoadedException;

    @Named("getPropertyDisplayableValue")
    static String getPropertyDisplayableValue(ModelObject modelObject, String source) throws NotLoadedException {
        return modelObject.getPropertyDisplayableValue(source);
    }
}
```

But, I noticed there are no workarounds something like this.  

I also don't want to manually implement mapping method something like [this](https://mapstruct.org/documentation/stable/reference/html/#adding-custom-methods). 

What I need is just specifying custom method to get values from source object.  

Am I missing something or this is not available with `MapStruct`?  

Thanks for helps.

**Comment by fjnajava** [comment]
Hi,

Check out the OTC framework - https://otcframework.org. It looks very promising and an awesome framework in my opinion.

Best thing I liked are - there are no interfaces, annotations or manual configurations - everything is extremely quick, simple and easy and hardly takes any time. And its performance is mind blowing.

**Comment by filiphr** [comment]
@wonsuc in the upcoming 1.6 release we are going to have `@TargetPropertyName` so doing something like:

```java
@Mapper
public interface ProjectMapper {
    ProjectMapper INSTANCE = Mappers.getMapper(ProjectMapper.class);

    @Mapping(target = "id", source = "prg0PlanId", qualifiedByName = "getPropertyDisplayableValue")
    ProjectDto toDto(ModelObject projectModel) throws NotLoadedException;

    @Named("getPropertyDisplayableValue")
    static String getPropertyDisplayableValue(ModelObject modelObject, @TargetPropertyName String source) throws NotLoadedException {
        return modelObject.getPropertyDisplayableValue(source);
    }
}
```

should be feasible with that.

This was done in #2688

**Comment by PPrydorozhnyi** [comment]
Hi @filiphr. It looks like `@TargetPropertyName` is applicable only for `@Condition` methods for now. so, https://github.com/mapstruct/mapstruct/issues/2688 doesn't help with a mapping case. Are there any plans to support other methods, please?

**Comment by filiphr** [comment]
Good point @PPrydorozhnyi. I'll reopen this to see how easy / difficult would be to support such a thing. We are also going to have `@SourcePropertyName` in 1.6.

**Comment by filiphr** [comment]
This is a slightly bigger change and I'd like to move features out of the 1.6 release so that we can release sooner. Therefore, moving this to 1.7.
